Restructure emulator to use effects

michaelpj opened the pull request:
Not done, do not review.

michaelpj commented:
This is ready for a look. I think it's quite nice. There are some awkward corners in actually interpreting the effects (I need to write some more comments), but overall it's okay, and I think the separation is really good.

I also think this could help us resolve the fact that `plutus-contract` library has a separate notion of traces. I think it could instead just have a new kind of effect, and an interpreter into `plutus-emulator`'s notion of traces. I think this would simplify a lot of things, e.g. the whole `EmulatorAction` dance.

I haven't pushed on e.g. making the notifications for wallets go via the node clients, which they probably should, since this PR is already pretty huge.

michaelpj commented:
Some test failures in `plutus-use-cases` that I haven't fixed yet too.

effectfully commented on plutus-emulator/src/Wallet/Emulator/MultiAgent.hs:
```diff
@@ -0,0 +1,220 @@
+            p1 :: Prism' [EmulatorEvent] [Wallet.WalletEvent]
+            p1 = below (walletEvent wallet)
+            p2 :: Prism' [EmulatorEvent] [NC.NodeClientEvent]
+            p2 = below (walletClientEvent wallet)
```
We might use a class here to get rid of the explicit prisms:

```haskell
class a <: s where
    subtype :: Prism' s a

instance NC.NodeClientEvent <: EmulatorEvent
```

michaelpj replied:
Don't typeclasses like this where you don't have a fundep behave quite badly?

Also note that you actually need to pass a parameter to get the prism, so I don't think that would work.

effectfully commented on plutus-emulator/src/Wallet/Emulator/MultiAgent.hs:
```diff
@@ -0,0 +1,220 @@
+    if value == total
+    then pure ()
+    else throwError $ GenericAssertion $ T.unwords ["Funds in wallet", tshow wallet, "were", tshow total, ". Expected:", tshow value]
+    where
```
That `if-then-else-where` without any indentation is hard to read to me.

michaelpj replied:
This stuff is just moved from the other file, but I agree.

effectfully commented on plutus-emulator/src/Wallet/Emulator/MultiAgent.hs:
```diff
@@ -0,0 +1,220 @@
+    emState <- get
+    if notElem txn (join $ emState ^. chainState . Chain.chainNewestFirst)
+        then throwError $ GenericAssertion $ "Txn not validated: " <> T.pack (show txn)
+        else pure ()
```
```haskell
when (notElem txn (join $ emState ^. chainState . Chain.chainNewestFirst)) $
    throwError $ GenericAssertion $ "Txn not validated: " <> T.pack (show txn)
```

effectfully commented on plutus-emulator/src/Wallet/Emulator/Wallet.hs:
```diff
@@ -0,0 +1,235 @@
+
+instance (Member (Error WAPI.WalletAPIError) effs) => E.MonadError WAPI.WalletAPIError (Eff effs) where
+    throwError = throwError
+    catchError = catchError
```
It's not a loop, because lhses come from `Control.Monad.Except` and rhses from `freer-simple`? Worth a short note, I think, because it looks like a loop.

michaelpj replied:
Yes, names inside typeclass instances are a bit funny.

effectfully commented on plutus-contract/plutus-contract.cabal:
```diff
@@ -83,6 +83,7 @@ library
         tasty -any,
         tasty-hunit -any,
         row-types -any,
+        freer-simple -any,
```
I'd try [`fused-effects`](http://hackage.haskell.org/package/fused-effects) first, because it supports higher-order effects (not sure if this is good, though) and seems to be a bit more widely used. On the other hand if a simple thing works, why complicate your life...

reactormonk replied:
I thought `polysemy` is all the rage right now.

michaelpj replied:
`freer-simple` was the first one I could get to build: `polysemy` got stuck since most of its dependencies aren't in the stack snapshot we're using. I didn't try `fused-effects` since `freer-simple` seems to do most of what I want and I don't care about fancy stuff like higher-order effects, but I can look at it. Fortunately they all seem to use similar styles of defining effects.

michaelpj commented on plutus-emulator/src/Control/Monad/Freer/Extras.hs:
```diff
@@ -0,0 +1,127 @@
+
+where effs is some *fixed* list of effects. You may then need to insert
+more effects *under* effs to interpret them in terms of. It turns out that
+inserting effects at the *end* of the list is tricky.
```
Sorry, I should have written down my other attempt, which was indeed to do that. However, I want to have a single type for the trace right now (otherwise it's *way* more refactoring), so you end up with `data SomeEffects effs a = SomeEffects (forall effs' . Members effs effs' => Eff effs a)` which is a real pain to work with.

michaelpj replied:
https://github.com/lexi-lambda/freer-simple/issues/30 describes it a bit more.

michaelpj commented on plutus-emulator/src/Wallet/Emulator/MultiAgent.hs:
```diff
@@ -0,0 +1,220 @@
+makeClassyPrisms ''AssertionError
+
+-- | This lets people use 'T.Text' as their error type.
+instance AsAssertionError T.Text where
```
Yeah, we should be able to get rid of a lot of this in the future since we can hopefully just have multiple error effects instead of the prism approach.

michaelpj commented on plutus-emulator/src/Wallet/Emulator/MultiAgent.hs:
```diff
@@ -0,0 +1,220 @@
+data MultiAgentEffect r where
+    -- | An direct action performed by a wallet. Usually represents a "user action", as it is
+    -- triggered externally.
+    WalletAction :: Wallet.Wallet -> Eff '[Wallet.WalletEffect, Error WAPI.WalletAPIError] r -> MultiAgentEffect r
```
Exactly: `MultiAgent` is the multi-party emulator, `Wallet` is just a single wallet.

michaelpj commented:
Everything seems to work now.

@krisajenkins I deleted the "decode an evaluation response" test. I'm not sure how useful it is and it's a *massive* pain every time we change any ledger types, because we just have to guess what the changes to the JSON-encoded version will be. I'm happy to keep it but only if we can auto-generate the output.

michaelpj commented:
I added a few more things:
- Moved the address map to the client, rather than the wallet, since that's what we think will hold it.
- Got rid of `Failure`, since it wasn't used any more.

j-mueller commented on plutus-wallet-api/ledger/Ledger/AddressMap.hs:
```diff
@@ -86,6 +88,10 @@ instance At AddressMap where
 
+-- | Get the funds available at a particular address.
+fundsAt :: Address -> Lens' AddressMap (Map TxOutRef TxOutTx)
+fundsAt addr = at addr . non mempty
```
nice

michaelpj commented:
Okay, let's go with this. More stuff in future.
